[HPB-153] [BE] 實作建立創作者方式

KR0924 opened the pull request:
# feat: add user management and shared database connection module
  - Implement artisan make:user command (interactive user creation with password encryption)
  - Add user SQL queries, Store layer, and sqlc generated code
  - Create pkg/database package for shared database connection functionality
  - Update Makefile to support artisan commands
  ---
  | 中文說明
  - 實作 artisan make:user 命令（互動式建立使用者，含密碼加密）
  - 新增 user SQL 查詢、Store 層與 sqlc 生成程式碼
  - 建立 pkg/database 套件，提供共用資料庫連線功能
  - 更新 Makefile 支援 artisan 命令
  
 ## CI
<img width="746" height="256" alt="image" src="https://github.com/user-attachments/assets/030d0c21-8208-4650-8837-9bb7602b20ac" />

 ## Test
Note: A /login route was implemented to test the user creation flow. However, to avoid conflicts with the authentication/login/logout functionality in [HPB-152], the implementation code  listed below is excluded from this commit.
註：為了測試使用者建立流程，實作了 /login 路由，但為避免與 [HPB-152] 的認證登入登出功能衝突，此次提交底下實作的程式碼不包含在此 commit 中。

 ```go
 // cmd/server/main.go
 package main

import (
	"context"
	"fmt"
	"log"
	"net/http"
	"os"

	"github.com/Goodidea-backend-camp/hpb-blog-backend/internal/api"
	"github.com/Goodidea-backend-camp/hpb-blog-backend/internal/db"
	"github.com/Goodidea-backend-camp/hpb-blog-backend/internal/store"
	"github.com/Goodidea-backend-camp/hpb-blog-backend/pkg/database"
	"github.com/gin-gonic/gin"
)

func run() error {
	dbpool, err := database.NewPool(context.Background())
	if err != nil {
		return err
	}
	defer dbpool.Close()

	log.Println("Successfully connected to the database!")

	// 初始化 DB Queries 和 Store
	queries := db.New(dbpool)
	postStore := store.NewPostStore(queries)
	userStore := store.NewUserStore(queries)

	// 初始化 Handler
	handler := api.NewHandler(postStore, userStore)

	// 初始化 Gin Router
	router := gin.Default()
	handler.RegisterRoutes(router)

	// 健康檢查路由
	router.GET("/healthz", func(c *gin.Context) {
		if err := dbpool.Ping(c); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"status": "Database connection is down"})
			return
		}
		c.JSON(http.StatusOK, gin.H{"status": "Database connection is healthy"})
	})

	backendPort := os.Getenv("BACKEND_PORT")
	if backendPort == "" {
		backendPort = "8080"
	}

	log.Printf("Starting server on :%s", backendPort)
	if err := router.Run(":" + backendPort); err != nil {
		return fmt.Errorf("server failed: %w", err)
	}

	return nil
}

func main() {
	if err := run(); err != nil {
		log.Fatal(err)
	}
}

 ```
 
 ```go
 // internal/api/handler.go
 package api

import (
	"github.com/Goodidea-backend-camp/hpb-blog-backend/internal/store"
	"github.com/gin-gonic/gin"
)

type Handler struct {
	store.PostStore
	userHandler *UserHandler
}

func NewHandler(postStore store.PostStore, userStore store.UserStore) *Handler {
	return &Handler{
		PostStore:   postStore,
		userHandler: NewUserHandler(userStore),
	}
}

func (h *Handler) RegisterRoutes(router *gin.Engine) {
	// User routes
	router.POST("/login", h.userHandler.Login)

	// Post routes
	router.POST("/posts", h.CreatePost)
	router.GET("/posts/:id", h.GetPost)
	router.GET("/posts", h.ListPosts)
	router.PUT("/posts/:id", h.UpdatePost)
	router.DELETE("/posts/:id", h.DeletePost)
}

 ```
 
  ```go
  // internal/api/user_handler.go
  package api

import (
	"database/sql"
	"errors"
	"net/http"

	"github.com/Goodidea-backend-camp/hpb-blog-backend/internal/store"
	"github.com/gin-gonic/gin"
	"golang.org/x/crypto/bcrypt"
)

type UserHandler struct {
	userStore store.UserStore
}

func NewUserHandler(userStore store.UserStore) *UserHandler {
	return &UserHandler{
		userStore: userStore,
	}
}

type LoginRequest struct {
	Username string `json:"username" binding:"required"`
	Password string `json:"password" binding:"required"`
}

type LoginResponse struct {
	ID       int32  `json:"id"`
	Username string `json:"username"`
	Message  string `json:"message"`
}

// Login 處理使用者登入
func (h *UserHandler) Login(c *gin.Context) {
	var req LoginRequest

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request"})
		return
	}

	// 查詢使用者
	user, err := h.userStore.GetUserByUsername(c.Request.Context(), req.Username)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid username or password"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
		return
	}

	// 驗證密碼
	err = bcrypt.CompareHashAndPassword([]byte(user.Password), []byte(req.Password))
	if err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid username or password"})
		return
	}

	// 登入成功
	c.JSON(http.StatusOK, LoginResponse{
		ID:       user.ID,
		Username: user.Username,
		Message:  "Login successful",
	})
}

 ```
 ### Result
<img width="693" height="311" alt="image" src="https://github.com/user-attachments/assets/159e7a1a-f960-4bf2-b609-78dfde1a61dc" />
<img width="829" height="137" alt="image" src="https://github.com/user-attachments/assets/95877ed7-5de9-4248-bea5-9745b83cea7c" />

KR0924 committed:
feat: add user management and shared database connection module

  - Implement artisan make:user command (interactive user creation with password encryption)
  - Add user SQL queries, Store layer, and sqlc generated code
  - Create pkg/database package for shared database connection functionality
  - Update Makefile to support artisan commands and environment checks

KR0924 committed:
chore: add dependencies for user authentication

  - Add golang.org/x/crypto for password hashing (bcrypt)
  - Add golang.org/x/term for secure password input

KR0924 committed:
style: fix lint error

JYu1999 commented on db/queries/user.sql:
看起來這些 query 似乎沒有全部都用到？提前實作出來的用意是什麼？

KR0924 replied:
目前的確只有用到 CreateUser 和 GetUserByUsername。

其他的查詢（GetUser、ListUsers、UpdateUser、DeleteUser）是預先準備的，方便後續實作使用者 CRUD 功能時可以直接使用。

JYu1999 commented on pkg/database/database.go:
想了解一下為什麼需要使用這個套件？

JYu1999 replied:
另外我也想知道為什麼選擇放在 pkg？是因為預期被外部專案使用嗎？

KR0924 replied:
主要是避免 cmd/server 和 cmd/artisan 的資料庫連線邏輯重複。

放在 pkg/ 是因為：
1. 這是基礎設施層級的共用功能
2. 根據 README 的架構說明，pkg/ 用於存放共用庫（如 logger、validator）
3. 資料庫連線邏輯屬於類似性質

如果覺得要放到 internal/ 更合適，我再調整。

JYu1999 commented on Makefile:
```diff
@@ -40,6 +40,23 @@ migrate-drop:
 sqlc:
 	@$(COMPOSE_CMD) exec backend sh -c 'sqlc generate'
 
+# ====================================================================================
+# Artisan Commands (CLI 工具指令)
+# ====================================================================================
+# artisan: 在 container 中執行 CLI 工具（推薦）
+# artisan-local: 在本機執行 CLI 工具（需要本機有 DATABASE_URL）
+#
+# 使用範例：
+#   make artisan make:user
+# ====================================================================================
```
這個我感覺應該寫在 Readme？

KR0924 replied:
已在[docs: add Artisan CLI usage documentation](https://github.com/Goodidea-backend-camp/hpb-blog-backend/pull/14/commits/a3f68a094d554ff0f70354dd3fb2a139cbf3329f) 調整

JYu1999 commented on internal/store/user_store.go:
```diff
@@ -0,0 +1,26 @@
+)
+
+type UserStore interface {
+	CreateUser(ctx context.Context, arg db.CreateUserParams) (db.User, error)
+	GetUser(ctx context.Context, id int32) (db.User, error)
+	GetUserByUsername(ctx context.Context, username string) (db.User, error)
+	ListUsers(ctx context.Context) ([]db.User, error)
+	UpdateUser(ctx context.Context, arg db.UpdateUserParams) (db.User, error)
+	DeleteUser(ctx context.Context, id int32) error
```
同 https://github.com/Goodidea-backend-camp/hpb-blog-backend/pull/14/files#r2576476204

KR0924 replied:
同https://github.com/Goodidea-backend-camp/hpb-blog-backend/pull/14#discussion_r2590965993

JYu1999 commented on Makefile:
```diff
@@ -40,6 +40,23 @@ migrate-drop:
+# Artisan Commands (CLI 工具指令)
+# ====================================================================================
+# artisan: 在 container 中執行 CLI 工具（推薦）
+# artisan-local: 在本機執行 CLI 工具（需要本機有 DATABASE_URL）
```
什麼情況下我們需要開放本機執行？

KR0924 replied:
因為到正式或測試站上的時候，我不希望有個指令可以在主機上執行然後就產生帳號
目前我看是prod環境是用編譯的，所以環境沒有go所以也沒辦法執行指令，所以這些指令都只會在本地開發時使用

JYu1999 commented on cmd/artisan/main.go:
```diff
@@ -0,0 +1,57 @@
+}
+
+func printHelp() {
+	fmt.Println("If you in Container Usage: go run cmd/artisan/main.go [command]")
+	fmt.Println("If you in Local Usage: make artisan [command]")
```
同 https://github.com/Goodidea-backend-camp/hpb-blog-backend/pull/14/files#r2576503039

KR0924 replied:
同 https://github.com/Goodidea-backend-camp/hpb-blog-backend/pull/14#discussion_r2590965619

KR0924 committed:
docs: add Artisan CLI usage documentation

  - Add Artisan CLI tool section to README
  - Document make:user command usage and examples
  - Include environment requirements and notes
